apps/src/lib/node/ledger/shims/abcipp_shim.rs

```diff
@@ -181,31 +213,101 @@ impl AbcippShim {
+    action_at_height: Option<Action>,
 }
 
-/// The ABCI tower service implementation sends and receives messages to and
-/// from the [`AbcippShim`] for requests from Tendermint.
-impl Service<Req> for AbciService {
-    type Error = BoxError;
-    type Future =
-        Pin<Box<dyn Future<Output = Result<Resp, BoxError>> + Send + 'static>>;
-    type Response = Resp;
-
-    fn poll_ready(
-        &mut self,
-        _cx: &mut Context<'_>,
-    ) -> Poll<Result<(), Self::Error>> {
-        // Nothing to check as the sender's channel is unbounded
-        Poll::Ready(Ok(()))
+impl AbciService {
+    /// Check if we are at a block height with a scheduled action.
+    /// If so, perform the action.
+    fn maybe_take_action(
+        action_at_height: Option<Action>,
+        check: CheckAction,
+        mut shutdown_recv: broadcast::Receiver<()>,
+    ) -> (bool, Option<<Self as Service<Req>>::Future>) {
+        let hght = match check {
+            CheckAction::AlreadySuspended => BlockHeight::from(u64::MAX),
+            CheckAction::Check(hght) => BlockHeight::from(hght as u64),
+            CheckAction::NoAction => BlockHeight::default(),
+        };
+        match action_at_height {
+            Some(Action::Suspend(height)) if height <= hght => {
+                if height == hght {
+                    tracing::info!(
+                        "Reached block height {}, suspending.",
+                        height
+                    );
+                    tracing::warn!(
+                        "\x1b[93mThis feature is intended for debugging \
+                         purposes. Note that on shutdown a spurious panic \
+                         message will be produced.\x1b[0m"
+                    )
+                }
+                (
+                    true,
+                    Some(Box::pin(
+                        async move {
+                            shutdown_recv.recv().await.unwrap();
+                            Err(BoxError::from(
+                                "Not all tendermint responses were processed. \
+                                 If the `--suspended` flag was passed, you \
+                                 may ignore this error.",
+                            ))
+                        }
+                        .boxed(),
+                    )),
+                )
+            }
+            Some(Action::Halt(height)) if height == hght => {
+                tracing::info!(
+                    "Reached block height {}, halting the chain.",
+                    height
+                );
+                (
+                    false,
+                    Some(Box::pin(
+                        async move {
+                            Err(BoxError::from(format!(
+                                "Reached block height {}, halting the chain.",
+                                height
+                            )))
+                        }
+                        .boxed(),
+                    )),
```

Comment:
this is boxing the future twice. you can get rid of the first `Box::pin()`.
```suggestion
                    Some(
                        async move {
                            Err(BoxError::from(format!(
                                "Reached block height {}, halting the chain.",
                                height
                            )))
                        }
                        .boxed(),
                    ),
```